### PR TITLE
Feature/search

### PR DESCRIPTION
**Pull request by bogdanStepanyuk**

### CODE REVIEW

File: app/src/main/java/com/uptech/example/vidme/ui/search/SearchActivity.java [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+import butterknife.Bind;
+
+/**
+ * Created by bogdan on 12.04.2016.
+ */
+public class SearchActivity extends BaseActivity{
```
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/ui/search/SearchActivity.java]: @bogdanStepanyuk please use proper formatting

File: app/src/main/java/com/uptech/example/vidme/ui/search/SearchActivity.java [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+    @Bind(R.id.toolbarSearch)
+    Toolbar toolbar;
+    @Bind(R.id.tabsSearch)
+    TabLayout tabLayout;
+    @Bind(R.id.viewpagerSearch)
```
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/ui/search/SearchActivity.java]: @bogdanStepanyuk please user snake_case naming convention for view ids

File: app/src/main/java/com/uptech/example/vidme/ui/search/SearchActivity.java [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    ViewPager viewPager;
+    ViewPagerAdapter adapter;
```
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/ui/search/SearchActivity.java]: @bogdanStepanyuk please use access modifiers (protected/private)

File: app/src/main/java/com/uptech/example/vidme/ui/search/SearchActivity.java [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+    private static final int VIDROS_FRAGMENT = 0;
+    private static final int USERS_FRAGMENT = 1;
```
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/ui/search/SearchActivity.java]: @bogdanStepanyuk please use correct fields order (link : http://stackoverflow.com/questions/4668218/are-there-any-java-method-ordering-conventions)

File: app/src/main/java/com/uptech/example/vidme/ui/search/SearchActivity.java [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+        setSupportActionBar(toolbar);
+        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
+        setupViewPager(viewPager);
+        tabLayout.setupWithViewPager(viewPager);
+    }
+
+    private void setupViewPager(ViewPager viewPager) {
+        adapter = new ViewPagerAdapter(getSupportFragmentManager());
+        adapter.addFragment(new VideosFragment(), "Videos");
+        adapter.addFragment(new UsersFragment(), "Users");
```
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/ui/search/SearchActivity.java]: @bogdanStepanyuk please use strings only from resources

File: app/src/main/java/com/uptech/example/vidme/ui/search/SearchActivity.java [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+        SearchView searchView = (SearchView) searchItem.getActionView();
+        searchView.setQueryHint("Search");
+        // need for focus when user open activity
+        searchView.setIconified(false);
+
+        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
+            @Override
+            public boolean onQueryTextSubmit(String query) {
+                VideosFragment videosFragment = (VideosFragment) adapter.getItemFragment(0);
+                if (!query.equals("")) {
```
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/ui/search/SearchActivity.java]: @bogdanStepanyuk please use `TextUtils.isEmpty()`

File: app/src/main/java/com/uptech/example/vidme/ui/search/SearchActivity.java [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+        public Fragment getItem(int position) {
+            return fragmentList.get(position);
+        }
+
+        public Fragment getItemFragment(int position) {
+            return fragmentList.get(position);
+        }
+
+
+        @Override
```
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/ui/search/SearchActivity.java]: @bogdanStepanyuk please update formatting (unnecessary spacing )

File: app/src/main/java/com/uptech/example/vidme/ui/search/SearchActivity.java [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+
+        SearchView searchView = (SearchView) searchItem.getActionView();
+        searchView.setQueryHint("Search");
+        // need for focus when user open activity
+        searchView.setIconified(false);
+
+        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
+            @Override
+            public boolean onQueryTextSubmit(String query) {
+                VideosFragment videosFragment = (VideosFragment) adapter.getItemFragment(0);
```
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/ui/search/SearchActivity.java]: @bogdanStepanyuk what we would do with Users fragment here ?

File: app/src/main/java/com/uptech/example/vidme/ui/search/fragments/VideosFragment.java [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+/**
+ * Created by bogdan on 12.04.2016.
+ */
+public class VideosFragment extends BaseVideoTabFragment {
+    private String query = "";
+
+    @Override
+    protected void loadVideos(int offset) {
+        if (!query.equals("")) {
```
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/ui/search/fragments/VideosFragment.java]: @bogdanStepanyuk please use `TextUtils.isEmpty()`

File: app/src/main/res/values/styles.xml [hunk 10]
```diff
@@ -8,6 +8,12 @@
         <item name="colorAccent">@color/colorAccent</item>
     </style>
 
+    <style name="SearchTheme" parent="Theme.AppCompat.Light.NoActionBar">
+        <item name="colorPrimary">@color/colorAccent</item>
+        <item name="colorPrimaryDark">@android:color/black</item>
+        <item name="colorAccent">@android:color/white</item>
```
Comment (AndriyBas) [inline app/src/main/res/values/styles.xml]: @bogdanStepanyuk please use local colour resources (not from `@android:`)

File: app/src/main/java/com/uptech/example/vidme/api/ApiService.java [hunk 11]
```diff
@@ -37,6 +37,9 @@
     @GET("videos/new")
     Call<VideoListResponse> getNewVideos(@Query("offset") int offset, @Query("limit") int limit);
 
+    @GET("videos/search")
+    Call<VideoListResponse> getSearchVideo(@Query("offset") int offset, @Query("query") String searchText);
```
Comment (Dima564) [inline app/src/main/java/com/uptech/example/vidme/api/ApiService.java]: We need to use Observable<VideoResponse>

File: app/src/main/java/com/uptech/example/vidme/ui/search/SearchActivity.java [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+public class SearchActivity extends BaseActivity{
+    @Bind(R.id.toolbarSearch)
+    Toolbar toolbar;
+    @Bind(R.id.tabsSearch)
+    TabLayout tabLayout;
+    @Bind(R.id.viewpagerSearch)
+    ViewPager viewPager;
+    ViewPagerAdapter adapter;
+
+    private static final int VIDROS_FRAGMENT = 0;
```
Comment (Dima564) [inline app/src/main/java/com/uptech/example/vidme/ui/search/SearchActivity.java]: VIDROSS!

File: app/src/main/java/com/uptech/example/vidme/adapters/UsersSearchListAdapter.java [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+        }
+        if (TextUtils.isEmpty(String.valueOf(user.getVideosViews()))) {
+            holder.views.setVisibility(View.GONE);
+        } else {
+            holder.views.setText(user.getUserName());
+        }
+        if (TextUtils.isEmpty(String.valueOf(user.getFollowerCount()))) {
+            holder.followers.setVisibility(View.GONE);
+        } else {
+            holder.followers.setText(user.getUserName());
```
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/adapters/UsersSearchListAdapter.java]: @bogdanStepanyuk there's an error in this code. We need to fix it, please remind about it tomorrow

File: app/src/main/java/com/uptech/example/vidme/adapters/VerticalSpaceItemDecoration.java [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+import android.graphics.Rect;
+import android.support.v7.widget.RecyclerView;
+import android.view.View;
+
+/**
+ * Created by bogdan on 13.04.2016.
+ */
+public class VerticalSpaceItemDecoration extends RecyclerView.ItemDecoration {
+
+    private final int mVerticalSpaceHeight;
```
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/adapters/VerticalSpaceItemDecoration.java]: @bogdanStepanyuk please don't use `m`-prefixed names

File: app/src/main/java/com/uptech/example/vidme/adapters/VideoSearchListAdapter.java [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+    @Override
+    public void onBindViewHolder(ViewHolder holder, int position) {
+        Video video = videoList.get(position);
+        Glide.with(context)
+                .load(video.getThumbnailUrl())
+                .into(holder.videoThumbnail);
+        if (TextUtils.isEmpty(video.getTitle())) {
+            holder.videoTitle.setVisibility(View.GONE);
+        } else {
+            holder.videoTitle.setText(video.getTitle());
```
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/adapters/VideoSearchListAdapter.java]: @bogdanStepanyuk also an error here

File: app/src/main/java/com/uptech/example/vidme/adapters/VideoSearchListAdapter.java [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+        }
+        holder.likesCount.setText(String.format("%d", video.getLikesCount()));
+        holder.viewsCount.setText(String.format("%d", video.getViewCount()));
```
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/adapters/VideoSearchListAdapter.java]: @bogdanStepanyuk too long, you don't need to write this to convert `int` to `String`. `String.valueOf()` works great

File: app/src/main/java/com/uptech/example/vidme/api/dto/UserResponse.java [hunk 17]
```diff
@@ -2,13 +2,15 @@
...
+import java.util.List;
+
 /**
  * Created by oleh on 31.03.16
  */
 public class UserResponse {
 //    private boolean status;
     private Auth auth;
-    private User user;
+    private List<User> users;
```
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/api/dto/UserResponse.java]: @bogdanStepanyuk you're modifying the class that's in use. It probably breaks other functionality. Isn't it ?

File: app/src/main/java/com/uptech/example/vidme/ui/search/SearchActivity.java [hunk 18]
```diff
@@ -0,0 +1,143 @@
...
+        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
+            @Override
+            public boolean onQueryTextSubmit(String query) {
+                VideosFragment videosFragment = adapter.getVideosFragment();
+                UsersFragment usersFragment = adapter.getUsersFragment();
+                if (!TextUtils.isEmpty(query)) {
+                    videosFragment.setQuery(query);
+                    videosFragment.loadVideos(0);
+                    usersFragment.setQuery(query);
+                    usersFragment.loadVideos(0);
```
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/ui/search/SearchActivity.java]: @bogdanStepanyuk we need to make this through `Callback`/`Listener` here

File: app/src/main/java/com/uptech/example/vidme/ui/search/fragments/UsersFragment.java [hunk 19]
```diff
@@ -0,0 +1,134 @@
...
+    public static String query = "";
+
+    @Inject
+    protected ApiService apiService;
+    @Bind(R.id.front_page_video_search_list)
+    protected RecyclerView videoList;
+
+    protected UsersSearchListAdapter adapter;
+    protected boolean isLoading;
+    private LinearLayoutManager layoutManager;
```
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/ui/search/fragments/UsersFragment.java]: @bogdanStepanyuk we don't need this field here

File: app/src/main/java/com/uptech/example/vidme/ui/search/fragments/VideosFragment.java [hunk 20]
```diff
@@ -0,0 +1,129 @@
...
+import butterknife.Bind;
+import retrofit2.Call;
+import retrofit2.Callback;
+import retrofit2.Response;
+
+/**
+ * Created by bogdan on 12.04.2016.
+ */
+public class VideosFragment extends BaseFragment {
+    protected final static int PAGE_SIZE = 30;
```
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/ui/search/fragments/VideosFragment.java]: @bogdanStepanyuk this constant should not be here

File: app/src/main/java/com/uptech/example/vidme/ui/search/fragments/UsersFragment.java [hunk 21]
```diff
@@ -0,0 +1,134 @@
...
+            }
+        });
+    }
+
+    public void loadVideos(int offset) {
+        Call<UserResponse> call = apiService.getUsers(offset, query);
+        call.enqueue(new Callback<UserResponse>() {
+            @Override
+            public void onResponse(Call<UserResponse> call, Response<UserResponse> response) {
+                handleResponse(response);
```
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/ui/search/fragments/UsersFragment.java]: @bogdanStepanyuk there's an error in this code

File: app/src/main/java/com/uptech/example/vidme/ui/search/fragments/VideosFragment.java [hunk 22]
```diff
@@ -0,0 +1,129 @@
...
+    private static String query = "";
+
+    @Inject
+    protected ApiService apiService;
+    @Bind(R.id.front_page_video_search_list)
+    protected RecyclerView videoList;
+
+    protected VideoSearchListAdapter adapter;
+    protected boolean isLoading;
+    private LinearLayoutManager layoutManager;
```
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/ui/search/fragments/VideosFragment.java]: @bogdanStepanyuk no need for this field

File: app/src/main/java/com/uptech/example/vidme/ui/search/fragments/UsersFragment.java [hunk 23]
```diff
@@ -0,0 +1,134 @@
...
+                int visibleItemCount = layoutManager.getChildCount();
+                int totalItemCount = layoutManager.getItemCount();
+                int firstVisibleItemPosition = layoutManager.findFirstVisibleItemPosition();
+
+                if (!isLoading) {
+                    if ((visibleItemCount + firstVisibleItemPosition) >= totalItemCount
+                            && firstVisibleItemPosition >= 0
+                            && totalItemCount >= PAGE_SIZE) {
+                        isLoading = true;
+                        loadVideos(totalItemCount);
```
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/ui/search/fragments/UsersFragment.java]: @bogdanStepanyuk this scroll listener is the same as in Users fragment. Let's create on method in order not to duplicate it

**Comment by AndriyBas** [comment]
@bogdanStepanyuk 1) we need to make icons smaller a bit (especially `eye`-icon)
2) we need to merge with develop

File: app/src/main/java/com/uptech/example/vidme/adapters/VerticalSpaceItemDecoration.java (hunk 14 above)
Comment (bogdanStepanyuk) [inline app/src/main/java/com/uptech/example/vidme/adapters/VerticalSpaceItemDecoration.java]: fixed

File: app/src/main/java/com/uptech/example/vidme/adapters/UsersSearchListAdapter.java (hunk 13 above)
Comment (bogdanStepanyuk) [inline app/src/main/java/com/uptech/example/vidme/adapters/UsersSearchListAdapter.java]: fixed

File: app/src/main/java/com/uptech/example/vidme/ui/search/fragments/UsersFragment.java (hunk 21 above)
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/ui/search/fragments/UsersFragment.java]: fixed

File: app/src/main/java/com/uptech/example/vidme/api/dto/UserResponse.java [hunk 24]
```diff
@@ -10,6 +10,7 @@
 public class UserResponse {
 //    private boolean status;
     private Auth auth;
+    private User user;
```
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/api/dto/UserResponse.java]: @bogdanStepanyuk pease create another class, it's not correct

File: app/src/main/java/com/uptech/example/vidme/ui/search/fragments/BaseSearchFragment.java [hunk 25]
```diff
@@ -0,0 +1,106 @@
...
+    protected ApiService apiService;
+    @Bind(R.id.front_page_video_search_list)
+    protected RecyclerView list;
+    @Bind(R.id.swipe_refresh_layout)
+    protected SwipeRefreshLayout swipe_refresh_layout;
+
+    protected boolean isLoading;
+
+    public static void setQuery(String query) {
+        BaseSearchFragment.query = query;
```
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/ui/search/fragments/BaseSearchFragment.java]: @bogdanStepanyuk please remove all static variables

File: app/src/main/java/com/uptech/example/vidme/ui/search/fragments/BaseSearchFragment.java [hunk 26]
```diff
@@ -0,0 +1,106 @@
...
+                        isLoading = true;
+                        loadVideos(totalItemCount);
+                    }
+                }
+            }
+        });
+    }
+
+    // need to Override
+    public abstract void loadVideos(int offset);
```
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/ui/search/fragments/BaseSearchFragment.java]: @bogdanStepanyuk please rename to `loadData`

File: app/src/main/java/com/uptech/example/vidme/ui/search/fragments/BaseSearchFragment.java [hunk 27]
```diff
@@ -0,0 +1,106 @@
...
+public abstract class BaseSearchFragment extends BaseFragment implements SwipeRefreshLayout.OnRefreshListener {
+    protected final static int PAGE_SIZE = 30;
+    protected static String query = "";
+
+    @Inject
+    protected ApiService apiService;
+    @Bind(R.id.front_page_video_search_list)
+    protected RecyclerView list;
+    @Bind(R.id.swipe_refresh_layout)
+    protected SwipeRefreshLayout swipe_refresh_layout;
```
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/ui/search/fragments/BaseSearchFragment.java]: @bogdanStepanyuk please use only camel case in Java classes

File: app/src/main/java/com/uptech/example/vidme/adapters/UsersSearchListAdapter.java [hunk 28]
```diff
@@ -0,0 +1,100 @@
...
+            holder.user_name.setVisibility(View.VISIBLE);
+            holder.user_name.setText(user.getUserName());
+        }
+        if (TextUtils.isEmpty(String.valueOf(user.getVideosViews()))) {
+            holder.views.setVisibility(View.GONE);
+        } else {
+            holder.views.setVisibility(View.VISIBLE);
+            holder.views.setText(user.getUserName());
+        }
+        if (TextUtils.isEmpty(String.valueOf(user.getFollowerCount()))) {
```
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/adapters/UsersSearchListAdapter.java]: @bogdanStepanyuk please check numbers for `null` or `0`, no need to convert to `String`

File: app/src/main/java/com/uptech/example/vidme/ui/front/BaseVideoTabFragment.java [hunk 29]
```diff
@@ -29,6 +31,7 @@
 public abstract class BaseVideoTabFragment extends BaseFragment {
 
     protected final static int PAGE_SIZE = 30;
+    protected static String query = "";
```
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/ui/front/BaseVideoTabFragment.java]: @bogdanStepanyuk `static` is evil, don't use it ....

File: app/src/main/java/com/uptech/example/vidme/ui/front/MainActivity.java [hunk 30]
```diff
@@ -105,8 +107,18 @@ public void startLoginActivity() {
...
+        Intent i = new Intent(this, SearchActivity.class);
+        startActivity(i);
+    }
+
+
     public class SectionsPagerAdapter extends FragmentPagerAdapter {
 
+        public static final int FEATURED_TAB = 0;
+        public static final int NEW_TAB = 1;
+        public static final int FOLLOWING_TAB = 2;
```
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/ui/front/MainActivity.java]: @bogdanStepanyuk looks like these variables are not used

File: app/src/main/java/com/uptech/example/vidme/ui/search/SearchActivity.java [hunk 31]
```diff
@@ -0,0 +1,151 @@
...
+ * Created by bogdan on 12.04.2016.
+ */
+public class SearchActivity extends BaseActivity {
+    @Bind(R.id.toolbar_search)
+    Toolbar toolbar;
+    @Bind(R.id.tabs_search)
+    TabLayout tab_layout;
+    @Bind(R.id.view_pager_search)
+    ViewPager viewPager;
+    ViewPagerAdapter adapter;
```
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/ui/search/SearchActivity.java]: @bogdanStepanyuk please use access modifiers

File: app/src/main/java/com/uptech/example/vidme/ui/search/fragments/BaseSearchFragment.java [hunk 32]
```diff
@@ -0,0 +1,106 @@
...
+
+    protected void onVideoLoadingFailure() {
+        isLoading = false;
+        swipe_refresh_layout.setRefreshing(false);
+        makeToast(R.string.front_loading_list_error);
+    }
+
+    @Override
+    public void onRefresh() {
+        loadVideos(0);
```
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/ui/search/fragments/BaseSearchFragment.java]: @bogdanStepanyuk we need to clear adapters here

File: app/src/main/java/com/uptech/example/vidme/ui/search/fragments/UsersFragment.java [hunk 33]
```diff
@@ -0,0 +1,60 @@
...
+
+/**
+ * Created by bogdan on 12.04.2016.
+ */
+public class UsersFragment extends BaseSearchFragment {
+    protected UsersSearchListAdapter adapter;
+
+    @Override
+    public void loadVideos(int offset) {
+        updateListAdapter();
```
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/ui/search/fragments/UsersFragment.java]: @bogdanStepanyuk we don't need to clear adapter here

File: app/src/main/res/layout/video_search_item.xml [hunk 34]
```diff
@@ -0,0 +1,107 @@
...
+                android:minHeight="80dp"
+                android:layout_margin="10dp"
+                android:layout_gravity="center"/>
+
+        </FrameLayout>
+
+        <FrameLayout
+            android:layout_width="match_parent"
+            android:layout_height="match_parent"
+            android:background="#212121"
```
Comment (AndriyBas) [inline app/src/main/res/layout/video_search_item.xml]: @bogdanStepanyuk please never use constants like that, create variables in resources and user only variables

File: app/src/main/res/layout/video_search_item.xml [hunk 35]
```diff
@@ -0,0 +1,107 @@
...
+                android:orientation="vertical">
+
+                <ImageView
+                    android:layout_width="30dp"
+                    android:layout_height="30dp"
+                    android:src="@drawable/expand_less"/>
+
+                <TextView
+                    android:id="@+id/likes_count"
+                    android:text="100500"
```
Comment (AndriyBas) [inline app/src/main/res/layout/video_search_item.xml]: @bogdanStepanyuk please update with `tools` namespace

**Pull request merged**